src/onnx/parse_reduce_op.cpp

```diff
@@ -31,66 +31,110 @@ namespace migraphx {
 
-instruction_ref parse_reduce_oper(const std::string& op_name,
-                                  const onnx_parser& parser,
-                                  onnx_parser::node_info info,
-                                  std::vector<instruction_ref> args)
+template <typename Derived>
+struct reduce_parser : op_parser<Derived>
 {
-    // default to reduce over all dimensions
-    std::vector<int64_t> axes;
-    if(args.size() == 2)
+    instruction_ref parse_reduce_oper(const std::string& op_name,
+                                      const onnx_parser& parser,
+                                      onnx_parser::node_info info,
+                                      std::vector<instruction_ref> args) const
     {
-        auto arg_axes = args.at(1)->eval();
-        check_arg_empty(arg_axes, "PARSE_" + op_name + ": cannot handle variable axes!");
-        axes.clear();
-        arg_axes.visit([&](auto s) { axes.assign(s.begin(), s.end()); });
-    }
-    else if(contains(info.attributes, "axes"))
-    {
-        axes.clear();
-        auto&& attr_axes = info.attributes["axes"].ints();
-        axes             = std::vector<int64_t>(attr_axes.begin(), attr_axes.end());
-    }
+        auto constant_axes = parse_constant_axes(args, info);
 
-    bool noop_with_empty_axes = false;
-    if(contains(info.attributes, "noop_with_empty_axes"))
-    {
-        noop_with_empty_axes = static_cast<bool>(
-            parser.parse_value(info.attributes.at("noop_with_empty_axes")).at<int>());
-    }
+        int noop_with_empty_axes =
+            parse_attribute<int>("noop_with_empty_axes", parser, info).value_or(0);
 
-    // empty axes behavior
-    if(axes.empty())
-    {
-        if(noop_with_empty_axes)
+        int keep_dims = parse_attribute<int>("keepdims", parser, info).value_or(1);
+
+        std::vector<int64_t> all_axes(args.front()->get_shape().ndim());
+        std::iota(all_axes.begin(), all_axes.end(), 0);
+
+        // Handle axes attribute, constant input axes, and missing both attribute and input cases
+        if(constant_axes.has_value())
         {
-            return args.at(0);
+            if(noop_with_empty_axes != 0 and constant_axes->empty())
+                return args[0];
+
+            if(noop_with_empty_axes == 0 and constant_axes->empty())
+                constant_axes = all_axes;
+
+            auto reduce =
+                info.add_instruction(make_op(op_name, {{"axes", *constant_axes}}), args[0]);
+
+            if(keep_dims == 0)
+                return info.add_instruction(make_op("squeeze", {{"axes", *constant_axes}}), reduce);
+
+            return reduce;
+        }
+
+        // Handle variable input axes
+        if(keep_dims == 0)
+            MIGRAPHX_THROW("Keepdims not supported with runtime provided axes");
+
+        // Empty axes attribute indicates to the operator to look for axes in the inputs
+        auto reduce_op = make_op(op_name, {{"axes", {}}});
+
+        if(noop_with_empty_axes != 0)
```

Comment:
This clause also needs to check that the axes is indeed empty.

Reply:
I'm sorry but I still fail to see why it should. Can you demonstrate with a particular situation for which it would be necessary, and how you see the code path for that situation, including the compute method of the operator.

Reply:
We cant check that the axes is empty because we dont know how many axes there are yet until runtime.

Reply:
(Had a discussion with Paul. The reason this logic works is because, our reduce_op has some default behavior when the axis is empty: to be an identity operator. Hence it is taken care of somewhere else. This code doesn't explain that.)
So please add a comment here to give a hint:
'If axes is empty, the default behavior of reduce_op is to be an identity operator"

Reply:
I added that default behavior, and that's what I was trying to explain. I've added the suggested comment. Thanks for reviewing.